language/tools/move-unit-test/src/test_runner.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    format_module_id,
+    test_reporter::{FailureReason, TestFailure, TestResults, TestStatistics},
+};
+use anyhow::Result;
+use colored::*;
+use crossbeam_channel::{bounded, Receiver, Sender};
+use move_binary_format::file_format::CompiledModule;
+use move_core_types::{
+    gas_schedule::{CostTable, GasAlgebra, GasCost, GasUnits},
+    identifier::IdentStr,
+    value::serialize_values,
+    vm_status::StatusCode,
+};
+use move_lang::unit_test::{ExpectedFailure, ModuleTestPlan, TestPlan};
+use move_vm_runtime::{logging::NoContextLog, move_vm::MoveVM};
+use move_vm_test_utils::InMemoryStorage;
+use move_vm_types::gas_schedule::{zero_cost_schedule, GasStatus};
+use std::{
+    io::Write,
+    marker::Send,
+    sync::{Arc, Mutex},
+    thread,
+};
+
+struct WorkerConfig<W: Write + Send> {
+    storage: InMemoryStorage,
+    cost_table: CostTable,
+    execution_bound: u64,
+    writer: Arc<Mutex<W>>,
+}
+
+#[derive(Debug)]
+pub struct TestRunner<W: Write + Send> {
+    tests: TestPlan,
+    workers: Vec<std::thread::JoinHandle<TestStatistics>>,
+    sender: Sender<ModuleTestPlan>,
+    writer: Arc<Mutex<W>>,
+}
+
+/// A gas schedule where every instruction has a cost of "1". This is used to bound execution of a
+/// test to a certain number of ticks.
+fn unit_cost_table() -> CostTable {
+    let mut cost_schedule = zero_cost_schedule();
+    cost_schedule.instruction_table.iter_mut().for_each(|cost| {
+        *cost = GasCost::new(1, 1);
+    });
+    cost_schedule.native_table.iter_mut().for_each(|cost| {
+        *cost = GasCost::new(1, 1);
+    });
+    cost_schedule
+}
+
+/// Setup storage state with the set of modules that will be needed for all tests
+fn setup_test_storage(modules: Vec<&CompiledModule>) -> Result<InMemoryStorage> {
+    let mut storage = InMemoryStorage::new();
+    for module in modules {
+        let module_id = module.self_id();
+        let mut module_bytes = Vec::new();
+        module.serialize(&mut module_bytes)?;
+        storage.publish_or_overwrite_module(module_id, module_bytes);
+    }
+
+    Ok(storage)
+}
+
+impl<W: 'static + Write + Send> TestRunner<W> {
+    pub fn new(
```

Comment:
Was a bit surprised to see so much custom worker logic. Judging by some of the errors, was this influenced by `cargo nextest` work?

I am definitely not the best to review this code. But can try a bit. For my own edification, why this versus something like rayon+par_iter ? Is there a lot of cross talk between threads?

Reply:
This is mainly to get all the types and lifetimes to work out so you can pass in a buffer for the testing output to be sent to. e.g., for the `main` we pass in stdout for this, but in the test runner, we pass in a string buffer and then write this to the `.exp` file. The types are a bit nasty, but I think it's worth it to allow callers to determine where they want the testing output to be rendered/captured.

Reply:
Yarp, I'm lost. We should chat about this.